### PR TITLE
vkaracic/enrollment-code-implementation: Add Enrollment Code products to support Bulk Purchasing

### PR DESCRIPTION
**Pull request by vkaracic**
This PR contains logic to create `Enrollment code` products and associate them with a seat.
Includes:
- a private method to create the `Enrollment code` product
- migrations for new `Enrollment code` product class and a new product attribute for the `Seat` product class
- settings for enabling automatic creation of enrollment codes
- tests

Part of https://openedx.atlassian.net/browse/WL-381

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @vkaracic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=712&repo=edx%2Fecommerce

File: ecommerce/extensions/api/serializers.py [hunk 1]
```diff
@@ -77,6 +78,10 @@ def get_value(self, obj):
             vouchers = obj.value.vouchers.all()
             serializer = VoucherSerializer(vouchers, many=True, context={'request': request})
             return serializer.data
+        if obj.attribute.name == ENROLLMENT_CODE:
+            request = self.context.get('request')
+            product_serializer = ProductSerializer(obj.value, many=False, context={'request': request})
```
Comment (vkaracic) [inline ecommerce/extensions/api/serializers.py]: Only parts that are interesting to us is the SKU, so instead of serializing the whole product, only send back the title and the SKU?

**Comment by mattdrayer** [comment]
Will need to squash commits

**Comment by vkaracic** [comment]
Of course. I'll squash it all in one commit once the reviews pass.

**Comment by mattdrayer** [comment]
Nice job separating this piece of the implementation from the original PR, @vkaracic!   Also I am seeing a couple of improvements included here that I had mentioned previously, which is great.  Looks like the branch is in need of an update/rebase, but otherwise I think we're good to move to review round #2 -- @clintonb, FYI

File: ecommerce/courses/models.py [hunk 2]
```diff
@@ -113,7 +114,8 @@ def _get_course_seat_name(self, certificate_type, id_verification_required):
         return name
 
     def create_or_update_seat(self, certificate_type, id_verification_required, price, partner,
-                              credit_provider=None, expires=None, credit_hours=None, remove_stale_modes=True):
+                              credit_provider=None, expires=None, credit_hours=None,
+                              remove_stale_modes=True, create_enrollment_code=False):
```
Comment (clintonb) [inline ecommerce/courses/models.py]: Enrollment code creation should be separate from seat creation.

File: ecommerce/courses/publishers.py [hunk 3]
```diff
@@ -29,11 +31,16 @@ def get_course_verification_deadline(self, course):
     def serialize_seat_for_commerce_api(self, seat):
         """ Serializes a course seat product to a dict that can be further serialized to JSON. """
         stock_record = seat.stockrecords.first()
+        try:
+            bulk_sku = StockRecord.objects.get(product=seat.attr.enrollment_code).partner_sku
```
Comment (clintonb) [inline ecommerce/courses/publishers.py]: Keep these products separate.

File: ecommerce/extensions/api/serializers.py [hunk 4]
```diff
@@ -77,6 +78,10 @@ def get_value(self, obj):
             vouchers = obj.value.vouchers.all()
             serializer = VoucherSerializer(vouchers, many=True, context={'request': request})
             return serializer.data
+        if obj.attribute.name == ENROLLMENT_CODE:
+            request = self.context.get('request')
```
Comment (clintonb) [inline ecommerce/extensions/api/serializers.py]: De-dupe. This is the same code form the clause above.

File: ecommerce/extensions/api/v2/tests/views/__init__.py [hunk 5]
```diff
@@ -49,7 +49,13 @@ def test_order_wrong_user(self):
 class ProductSerializerMixin(object):
     def serialize_product(self, product):
         """ Serializes a Product to a Python dict. """
-        attribute_values = [{'name': av.attribute.name, 'value': av.value} for av in product.attribute_values.all()]
+        attribute_values = [
+            {
+                'name': av.attribute.name,
+                'value': av.value if av.attribute.name != ENROLLMENT_CODE else self.serialize_product(av.value),
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/__init__.py]: What is going on here?

File: ecommerce/extensions/api/v2/tests/views/test_products.py [hunk 6]
```diff
@@ -87,14 +87,23 @@ def test_list_for_course(self):
...
         path = reverse('api:v2:course-product-list', kwargs={'parent_lookup_course_id': self.course.id})
         response = self.client.get(path)
         self.assertEqual(response.status_code, 200)
         results = [self.serialize_product(p) for p in self.course.products.all()]
         expected = {'count': 2, 'next': None, 'previous': None, 'results': results}
         self.assertDictEqual(json.loads(response.content), expected)
 
+    def test_enrollment_code_serialization(self):
+        """ Verify the view supports listing seats with enrollment codes. """
+        course = Course.objects.create(id='edX/DemoX/EC', name='EC Course')
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_products.py]: Use `CourseFactory`.

File: ecommerce/extensions/catalogue/migrations/0017_enrollment_code_product_class.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+ProductClass = get_model('catalogue', 'ProductClass')
+
+
+def create_enrollment_code_product_class(apps, schema_editor):
+    """Create a Enrollment code product class."""
+
+    enrollment_code = ProductClass.objects.create(
+        track_stock=False,
+        requires_shipping=False,
+        name='Enrollment code',
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/migrations/0017_enrollment_code_product_class.py]: Enrollment Code

File: ecommerce/extensions/catalogue/migrations/0017_enrollment_code_product_class.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        product_class=enrollment_code,
+        name='Course Key',
+        code='course_key',
+        type='text',
+        required=True
+    )
+
+    ProductAttribute.objects.create(
+        product_class=enrollment_code,
+        name='Seat type',
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/migrations/0017_enrollment_code_product_class.py]: Seat Type

File: ecommerce/extensions/catalogue/migrations/0017_enrollment_code_product_class.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        code='seat_type',
+        type='text',
+        required=True
+    )
+
+    seat_product_class = ProductClass.objects.get(slug='seat')
+    ProductAttribute.objects.create(
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/migrations/0017_enrollment_code_product_class.py]: Why is this necessary?

File: ecommerce/extensions/catalogue/utils.py [hunk 10]
```diff
@@ -23,6 +25,12 @@ def generate_sku(product, partner, **kwargs):
             unicode(catalog.id),
             str(partner.id)
         ))
+    elif product.product_class and product.product_class.name == ENROLLMENT_CODE:
+        _hash = ' '.join((
+            getattr(product.attr, 'course_key', ''),
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/utils.py:33]: `product.attr.course_key`

File: ecommerce/extensions/catalogue/utils.py [hunk 11]
```diff
@@ -23,6 +25,12 @@ def generate_sku(product, partner, **kwargs):
...
+            getattr(product.attr, 'seat_type', ''),
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/utils.py:34]: `product.attr.seat_type`

File: ecommerce/extensions/catalogue/utils.py [hunk 12]
```diff
@@ -23,6 +25,12 @@ def generate_sku(product, partner, **kwargs):
             unicode(catalog.id),
             str(partner.id)
         ))
+    elif product.product_class and product.product_class.name == ENROLLMENT_CODE:
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/utils.py]: De-dupe the `product.product_class` check.

File: ecommerce/settings/base.py [hunk 13]
```diff
@@ -516,3 +517,7 @@
 THEME_CACHE_TIMEOUT = 30 * 60
 
 # End Theme settings
+
+# Enrollment Code product settings
+AUTOGENERATE_ENROLLMENT_CODES = True  # Default behavior is to create bulk enrollment codes for seats
```
Comment (clintonb) [inline ecommerce/settings/base.py]: If this is supposed to be a feature flag, use [`waffle`](http://waffle.readthedocs.io/en/v0.11/).

File: ecommerce/settings/base.py [hunk 14]
```diff
@@ -516,3 +517,7 @@
...
+ENROLLMENT_CODE_VOUCHER_END_DATE = datetime.datetime.now() + datetime.timedelta(days=3650)
```
Comment (clintonb) [inline ecommerce/settings/base.py]: This is not used, and should be removed. Why the 10 year expiration? Just make the codes infinite.

File: ecommerce/courses/tests/test_models.py [hunk 15]
```diff
@@ -129,10 +134,27 @@ def test_create_or_update_seat(self):
...
+        )
+        self.assert_course_seat_valid(
+            seat, course, certificate_type, id_verification_required,
+            price, create_enrollment_code=create_enrollment_code
+        )
+
     def test_create_credit_seats(self):
         """Verify that the model's seat creation method allows the creation of multiple credit seats."""
         course = Course.objects.create(id='a/b/c', name='Test Course')
-        credit_data = {'MIT': 2, 'Harvard': 0.5}
```
Comment (clintonb) [inline ecommerce/courses/tests/test_models.py:135]: Why the change?

File: ecommerce/courses/models.py [hunk 16]
```diff
@@ -227,3 +231,47 @@ def create_or_update_seat(self, certificate_type, id_verification_required, pric
             ).delete()
 
         return seat
+
+    def _create_enrollment_code(self, course_id, certificate_type, partner, price):
+        """
+        Creates an enrollment code product for the created seat, setting the course ID
```
Comment (clintonb) [inline ecommerce/courses/models.py]: Remove the references to seats.

File: ecommerce/courses/models.py [hunk 17]
```diff
@@ -227,3 +231,47 @@ def create_or_update_seat(self, certificate_type, id_verification_required, pric
...
+            certificate_type (str): Seat type.
+            partner (Partner): Seat provider set in the stock record.
+            price (int): Price of the seat.
+
+        Returns:
+            Enrollment code product.
+        """
+
+        enrollment_code_product_class = ProductClass.objects.get(slug='enrollment_code')
+        title = '{} ({}) Enrollment Code'.format(self.name, certificate_type)
```
Comment (clintonb) [inline ecommerce/courses/models.py]: `Enrollment code for {seat_type} seat in {course_name}`

File: ecommerce/courses/models.py [hunk 18]
```diff
@@ -227,3 +231,47 @@ def create_or_update_seat(self, certificate_type, id_verification_required, pric
...
+        enrollment_code, created = Product.objects.get_or_create(
+            title=title,
```
Comment (clintonb) [inline ecommerce/courses/models.py]: Title should not be part of `get_or_create`. Try the following:

``` python
enrollment_code, created = Product.objects.get_or_create(product_class=enrollment_code_product_class, course=self)
```

Comment (clintonb) [inline ecommerce/courses/models.py]: You'll also want to filter by the `seat_type` attribute.

File: ecommerce/courses/models.py [hunk 19]
```diff
@@ -227,3 +231,47 @@ def create_or_update_seat(self, certificate_type, id_verification_required, pric
...
+        if created:
+            enrollment_code.attr.course_key = course_id
+            enrollment_code.attr.seat_type = certificate_type
+            enrollment_code.save()
+
+        try:
+            stock_record = StockRecord.objects.get(product=enrollment_code, partner=partner)
+        except StockRecord.DoesNotExist:
+            enrollment_code_sku = generate_sku(enrollment_code, partner)
+            stock_record = StockRecord.objects.create(
```
Comment (clintonb) [inline ecommerce/courses/models.py]: Save the database call, and just call the constructor.

File: ecommerce/core/constants.py [hunk 20]
```diff
@@ -7,6 +7,9 @@
 COURSE_ID_REGEX = r'[^/+]+(/|\+)[^/+]+(/|\+)[^/]+'
 COURSE_ID_PATTERN = r'(?P<course_id>{})'.format(COURSE_ID_REGEX)
 
+# Enrollment code product class name.
+ENROLLMENT_CODE = 'Enrollment code'
```
Comment (clintonb) [inline ecommerce/core/constants.py]: `ENROLLMENT_CODE_PRODUCT_CLASS_NAME`

If you're going to have this constant, use it consistently. There are still places where the name is hard-coded.

File: ecommerce/courses/tests/test_models.py (hunk 15 above)
Comment (vkaracic) [inline ecommerce/courses/tests/test_models.py:135]: I was getting an assertion error (0 != 0.5). I don't know how this ever worked since the `credit_data` attribute is of type integer: https://github.com/edx/ecommerce/blob/master/ecommerce/extensions/catalogue/migrations/0009_credit_hours_attr.py#L19

File: ecommerce/extensions/api/v2/tests/views/__init__.py (hunk 5 above)
Comment (vkaracic) [inline ecommerce/extensions/api/v2/tests/views/__init__.py]: The `enrollment_code` attribute is an entity and needs to be serialized.

File: ecommerce/extensions/catalogue/migrations/0017_enrollment_code_product_class.py (hunk 9 above)
Comment (vkaracic) [inline ecommerce/extensions/catalogue/migrations/0017_enrollment_code_product_class.py]: The plan was to include it to the seat so the publisher can send the data to LMS. I'll remove it when I decouple everything.

File: ecommerce/settings/base.py (hunk 14 above)
Comment (vkaracic) [inline ecommerce/settings/base.py]: It is used in the fulfillment module PR: https://github.com/edx/ecommerce/pull/719/files#diff-c9f38bb98fd490cc1127bb1cbdd272a3R416 I missed this one when I separated the code in multiple PR's. I can set this to be `datetime.max`.
@mattdrayer you mentioned that this could be changed at some point therefor we added it as a setting, but I'm not sure if this should be changeable?

File: ecommerce/courses/models.py (hunk 16 above)
Comment (vkaracic) [inline ecommerce/courses/models.py]: How so? Enrollment codes should be created for seats, not courses.

File: ecommerce/extensions/catalogue/utils.py (hunk 10 above)
Comment (vkaracic) [inline ecommerce/extensions/catalogue/utils.py:33]: So in these cases it's better that an `AttributeError` is raised?

File: ecommerce/courses/publishers.py (hunk 3 above)
Comment (vkaracic) [inline ecommerce/courses/publishers.py]: Why so?

File: ecommerce/settings/base.py [hunk 21]
```diff
@@ -1,4 +1,5 @@
 """Common settings and globals."""
+import datetime
```
Comment (clintonb) [inline ecommerce/settings/base.py]: Remove.

File: ecommerce/courses/models.py [hunk 22]
```diff
@@ -227,3 +230,57 @@ def create_or_update_seat(self, certificate_type, id_verification_required, pric
...
         return seat
+
+    def _create_or_update_enrollment_code(self, course_id, certificate_type, partner, price):
+        """
+        Creates an enrollment code product for the created seat, setting the course ID
+        and seat type in it's attributes. Also creates a purchasable stock record for the
+        enrollment code product.
+
+        Args:
+            course_id (str): ID of the seat's course.
```
Comment (clintonb) [inline ecommerce/courses/models.py]: This is not necessary. Use `self.id`.

File: ecommerce/courses/models.py [hunk 23]
```diff
@@ -227,3 +230,57 @@ def create_or_update_seat(self, certificate_type, id_verification_required, pric
...
+            certificate_type (str): Seat type.
```
Comment (clintonb) [inline ecommerce/courses/models.py]: This argument should be named `seat_type`.

File: ecommerce/courses/models.py [hunk 24]
```diff
@@ -227,3 +230,57 @@ def create_or_update_seat(self, certificate_type, id_verification_required, pric
...
+        Returns:
+            Enrollment code product.
+        """
+
+        enrollment_code_product_class = ProductClass.objects.get(slug='enrollment_code')
+        try:
+            enrollment_code = Product.objects.filter(
+                product_class=enrollment_code_product_class,
+                course=self
+            ).get(
```
Comment (clintonb) [inline ecommerce/courses/models.py]: Why make a second call here?

File: ecommerce/courses/models.py [hunk 25]
```diff
@@ -227,3 +230,57 @@ def create_or_update_seat(self, certificate_type, id_verification_required, pric
...
+            partner (Partner): Seat provider set in the stock record.
+            price (int): Price of the seat.
+
+        Returns:
+            Enrollment code product.
+        """
+
+        enrollment_code_product_class = ProductClass.objects.get(slug='enrollment_code')
+        try:
+            enrollment_code = Product.objects.filter(
```
Comment (clintonb) [inline ecommerce/courses/models.py]: Similar to how we have `self.seat_products`, create/use `self.enrollment_code_products`.

File: ecommerce/courses/models.py [hunk 26]
```diff
@@ -227,3 +230,57 @@ def create_or_update_seat(self, certificate_type, id_verification_required, pric
...
+            ).get(
+                attributes__name='seat_type',
+                attribute_values__value_text=certificate_type
+            )
+        except Product.DoesNotExist:
+            title = 'Enrollment code for {seat_type} seat in {course_name}'.format(
+                seat_type=certificate_type,
+                course_name=self.name
+            )
+            enrollment_code = Product()
```
Comment (clintonb) [inline ecommerce/courses/models.py]: This can be a single call. No need to break this into four lines.

File: ecommerce/courses/models.py [hunk 27]
```diff
@@ -227,3 +230,57 @@ def create_or_update_seat(self, certificate_type, id_verification_required, pric
...
+
+        enrollment_code.attr.course_key = course_id
+        enrollment_code.attr.seat_type = certificate_type
+        enrollment_code.save()
+
+        try:
+            stock_record = StockRecord.objects.get(product=enrollment_code, partner=partner)
+        except StockRecord.DoesNotExist:
+            enrollment_code_sku = generate_sku(enrollment_code, partner)
+            stock_record = StockRecord()
```
Comment (clintonb) [inline ecommerce/courses/models.py]: Same as above. Do this in a single line rather than splitting it across four calls.

File: ecommerce/courses/publishers.py [hunk 28]
```diff
@@ -29,11 +33,23 @@ def get_course_verification_deadline(self, course):
     def serialize_seat_for_commerce_api(self, seat):
         """ Serializes a course seat product to a dict that can be further serialized to JSON. """
         stock_record = seat.stockrecords.first()
+        try:
+            enrollment_code = Product.objects.filter(
```
Comment (clintonb) [inline ecommerce/courses/publishers.py]:
``` python
enrollment_code = seat.course.enrollment_code_products.filter(attributes__name='seat_type', attribute_values__value_text=seat.attr.certificate_type)
```

File: ecommerce/courses/publishers.py [hunk 29]
```diff
@@ -29,11 +33,23 @@ def get_course_verification_deadline(self, course):
...
+                product_class__name=ENROLLMENT_CODE_PRODUCT_CLASS_NAME,
+                course__id=seat.attr.course_key
+            ).get(
+                attributes__name='seat_type',
+                attribute_values__value_text=seat.attr.certificate_type
+            )
+            bulk_sku = StockRecord.objects.get(product=enrollment_code).partner_sku
+        except (StockRecord.DoesNotExist, Product.DoesNotExist):
```
Comment (clintonb) [inline ecommerce/courses/publishers.py]: If you get `StockRecord.DoesNotExist`, you have a serious problem with the product catalogue. At a minimum, the error should be logged if you're catching it. However, I wouldn't catch it. Let it bubble up, forcing someone to investigate.

File: ecommerce/courses/tests/test_models.py [hunk 30]
```diff
@@ -129,10 +133,28 @@ def test_create_or_update_seat(self):
         seat = course.seat_products[0]
         self.assert_course_seat_valid(seat, course, certificate_type, id_verification_required, price)
 
+    def test_create_seat_with_enrollment_code(self):
+        """Verify an enrollment code product is created."""
+        course = CourseFactory()
+        certificate_type = 'verified'
+        price = 5
+        switch, __ = Switch.objects.get_or_create(name=settings.ENROLLMENT_CODE_SWITCH, defaults={'active': True})
```
Comment (clintonb) [inline ecommerce/courses/tests/test_models.py]: Use `toggle_switch`.

File: ecommerce/courses/tests/test_models.py [hunk 31]
```diff
@@ -129,10 +133,28 @@ def test_create_or_update_seat(self):
...
+        switch.active = True
+        switch.save()
+        course.create_or_update_seat(certificate_type, True, price, self.partner)
+
+        enrollment_code = Product.objects.get(product_class__name=ENROLLMENT_CODE_PRODUCT_CLASS_NAME)
+        self.assertEqual(enrollment_code.attr.course_key, course.id)
+        self.assertEqual(enrollment_code.attr.seat_type, certificate_type)
+        self.assertEqual(enrollment_code.attr.seat_type, certificate_type)
```
Comment (clintonb) [inline ecommerce/courses/tests/test_models.py]: Delete

File: ecommerce/courses/tests/test_models.py [hunk 32]
```diff
@@ -129,10 +133,28 @@ def test_create_or_update_seat(self):
         seat = course.seat_products[0]
         self.assert_course_seat_valid(seat, course, certificate_type, id_verification_required, price)
 
+    def test_create_seat_with_enrollment_code(self):
+        """Verify an enrollment code product is created."""
+        course = CourseFactory()
+        certificate_type = 'verified'
```
Comment (clintonb) [inline ecommerce/courses/tests/test_models.py]: `seat_type`

File: ecommerce/courses/tests/test_models.py [hunk 33]
```diff
@@ -129,10 +133,28 @@ def test_create_or_update_seat(self):
...
+        switch.save()
+        course.create_or_update_seat(certificate_type, True, price, self.partner)
+
+        enrollment_code = Product.objects.get(product_class__name=ENROLLMENT_CODE_PRODUCT_CLASS_NAME)
+        self.assertEqual(enrollment_code.attr.course_key, course.id)
+        self.assertEqual(enrollment_code.attr.seat_type, certificate_type)
+        self.assertEqual(enrollment_code.attr.seat_type, certificate_type)
+
+        stock_record = StockRecord.objects.get(product=enrollment_code)
+        self.assertEqual(stock_record.price_excl_tax, price)
```
Comment (clintonb) [inline ecommerce/courses/tests/test_models.py:149]: What about currency and partner?

File: ecommerce/extensions/api/v2/tests/views/test_products.py [hunk 34]
```diff
@@ -87,7 +87,6 @@ def test_list_for_course(self):
         # Create another course and seat to confirm filtering.
         other_course = Course.objects.create(id='edX/DemoX/XYZ', name='Test Course 2')
         other_course.create_or_update_seat('honor', False, 0, self.partner)
-
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_products.py:90]: Remove.

File: ecommerce/extensions/catalogue/tests/mixins.py [hunk 35]
```diff
@@ -37,6 +40,23 @@ def seat_product_class(self):
                 ('credit_provider', 'text'),
                 ('id_verification_required', 'boolean'),
                 ('credit_hours', 'integer'),
+                ('enrollment_code', 'entity')
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/tests/mixins.py]: This is no longer necessary.

File: ecommerce/extensions/catalogue/tests/mixins.py [hunk 36]
```diff
@@ -37,6 +40,23 @@ def seat_product_class(self):
...
+            )
+
+            for code, attr_type in attributes:
+                factories.ProductAttributeFactory(code=code, name=code, product_class=pc, type=attr_type)
+
+        return pc
+
+    @property
+    def enrollment_code_product_class(self):
+        defaults = {'requires_shipping': False, 'track_stock': False, 'name': ENROLLMENT_CODE_PRODUCT_CLASS_NAME}
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/tests/mixins.py]: This shares code with `seat_product_class`. De-dup with an intermediary function:

``` python
def _create_product_class(self, class_name, slug, attributes):
  ...
```

File: ecommerce/extensions/catalogue/utils.py [hunk 37]
```diff
@@ -14,15 +16,20 @@ def generate_sku(product, partner, **kwargs):
...
-    # Note: This currently supports coupons and seats.
-    # A new product type can be added via a new else if block.
-    if product.product_class and product.product_class.name == 'Coupon':
-        catalog = kwargs.get('catalog', '')
-        _hash = ' '.join((
-            unicode(product.id),
-            unicode(catalog.id),
-            str(partner.id)
-        ))
+    if product.product_class:
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/utils.py]: I just realized that this makes no sense. Every product has a product class. Please improve this method:

``` python
def generate_sku(product, partner, **kwargs):
  product_class = product.get_product_class()

  if not product_class:
    # TODO Raise an error
    pass

  if product_class.slug == 'coupon':
    pass
  elif product_class.slug == 'seat':
    pass
  elif product_class.slug == 'enrollment_code':
    pass
  else:
    # TODO Raise an error because we have an unexpected product class.
```

**Comment by clintonb** [comment]
I'm almost ready to approve, but I need some clarification on your plan to push this data from Otto to LMS. What is the plan?

**Comment by mattdrayer** [comment]
@clintonb -- for new seat products, we will be creating a corresponding enrollment code product and pushing the "bulk sku" along with the rest of the publish payload.  For existing seat products we will be introducing a management command to create enrollment codes and push the bulk sku to LMS as part of that workflow.

**Comment by clintonb** [comment]
@mattdrayer where's the LMS PR?

File: ecommerce/courses/models.py (hunk 24 above)
Comment (vkaracic) [inline ecommerce/courses/models.py]: Is it possible for a course to have both `professional` and `verified` course modes?

Comment (clintonb) [inline ecommerce/courses/models.py]: No.

**Comment by mattdrayer** [comment]
https://github.com/edx/edx-platform/pull/12315

File: ecommerce/courses/publishers.py (hunk 28 above)
Comment (vkaracic) [inline ecommerce/courses/publishers.py]: If a course can have one course mode, the course should only have one enrollment code, therefor I don't see the need for a `filter` here?

File: ecommerce/core/constants.py [hunk 38]
```diff
@@ -7,6 +7,9 @@
 COURSE_ID_REGEX = r'[^/+]+(/|\+)[^/+]+(/|\+)[^/]+'
 COURSE_ID_PATTERN = r'(?P<course_id>{})'.format(COURSE_ID_REGEX)
 
+# Enrollment code product class name.
+ENROLLMENT_CODE_PRODUCT_CLASS_NAME = 'Enrollment code'
```
Comment (clintonb) [inline ecommerce/core/constants.py]: Capital `C`

File: ecommerce/settings/base.py [hunk 39]
```diff
@@ -516,3 +516,6 @@
 THEME_CACHE_TIMEOUT = 30 * 60
 
 # End Theme settings
+
+# Enrollment codes settings
+ENROLLMENT_CODE_SWITCH = 'create_enrollment_codes'
```
Comment (clintonb) [inline ecommerce/settings/base.py]: This should not be a setting. It's constant.

File: ecommerce/extensions/catalogue/tests/test_utils.py [hunk 40]
```diff
@@ -27,6 +29,17 @@ def setUp(self):
         self.course.create_or_update_seat('verified', False, 0, self.partner)
         self.catalog = Catalog.objects.create(name='Test', partner_id=self.partner.id)
 
+    def test_generate_sku_attribute_error(self):
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/tests/test_utils.py]: `test_generate_sku_with_missing_product_class`

File: ecommerce/extensions/catalogue/tests/test_utils.py [hunk 41]
```diff
@@ -27,6 +29,17 @@ def setUp(self):
...
+        """Verify the method raises an exception if the product class is missing."""
+        with self.assertRaises(AttributeError):
+            generate_sku(Product(), self.partner)
+
+    def test_generate_sku_error(self):
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/tests/test_utils.py]: `test_generate_sku_with_unexpected_product_class`

File: ecommerce/extensions/api/serializers.py [hunk 42]
```diff
@@ -12,7 +12,7 @@
 from rest_framework.reverse import reverse
 import waffle
 
-from ecommerce.core.constants import ISO_8601_FORMAT, COURSE_ID_REGEX
```
Comment (clintonb) [inline ecommerce/extensions/api/serializers.py:15]: None of the changes in this file are necessary.

File: ecommerce/courses/models.py [hunk 43]
```diff
@@ -182,6 +185,9 @@ def create_or_update_seat(self, certificate_type, id_verification_required, pric
         seat.attr.course_key = course_id
         seat.attr.id_verification_required = id_verification_required
 
+        if waffle.switch_is_active('create_enrollment_codes'):
```
Comment (mattdrayer) [inline ecommerce/courses/models.py]: AH, I need to use the switch constant here -- will update shortly

File: ecommerce/courses/publishers.py (hunk 28 above)
Comment (clintonb) [inline ecommerce/courses/publishers.py]: Either way, you only need one database call.

File: ecommerce/courses/models.py [hunk 44]
```diff
@@ -182,6 +185,9 @@ def create_or_update_seat(self, certificate_type, id_verification_required, pric
         seat.attr.course_key = course_id
         seat.attr.id_verification_required = id_verification_required
 
+        if waffle.switch_is_active(ENROLLMENT_CODE_SWITCH):
+            self._create_or_update_enrollment_code(certificate_type, partner, price)
```
Comment (clintonb) [inline ecommerce/courses/models.py:193]: How are you filtering for non-paid seats?

File: ecommerce/courses/models.py [hunk 45]
```diff
@@ -227,3 +231,59 @@ def create_or_update_seat(self, certificate_type, id_verification_required, pric
...
+    def _create_or_update_enrollment_code(self, seat_type, partner, price):
+        """
+        Creates an enrollment code product for the created seat, setting the course ID
+        and seat type in it's attributes. Also creates a purchasable stock record for the
+        enrollment code product.
+
+        Args:
+            seat_type (str): Seat type.
+            partner (Partner): Seat provider set in the stock record.
+            price (int): Price of the seat.
```
Comment (clintonb) [inline ecommerce/courses/models.py]: Prices are of type `Decimal`.

File: ecommerce/courses/models.py [hunk 46]
```diff
@@ -227,3 +231,59 @@ def create_or_update_seat(self, certificate_type, id_verification_required, pric
...
+    def enrollment_code_product(self):
+        """ Returns an enrollment code Product related to this course. """
+        return Product.objects.get(
+            product_class__name=ENROLLMENT_CODE_PRODUCT_CLASS_NAME,
+            course=self
+        )
+
+    def _create_or_update_enrollment_code(self, seat_type, partner, price):
+        """
+        Creates an enrollment code product for the created seat, setting the course ID
```
Comment (clintonb) [inline ecommerce/courses/models.py]: `Creates an enrollment code product, and stock record, ...` Drop the last sentence.

File: ecommerce/courses/models.py [hunk 47]
```diff
@@ -227,3 +231,59 @@ def create_or_update_seat(self, certificate_type, id_verification_required, pric
...
+                title=title,
+                product_class=enrollment_code_product_class,
+                course=self
+            )
+        enrollment_code.attr.course_key = self.id
+        enrollment_code.attr.seat_type = seat_type
+        enrollment_code.save()
+
+        try:
+            stock_record = StockRecord.objects.get(product=enrollment_code, partner=partner)
```
Comment (clintonb) [inline ecommerce/courses/models.py:279]: FYI: Django also supports [`update_or_create`](https://docs.djangoproject.com/en/1.8/ref/models/querysets/#update-or-create).

File: ecommerce/courses/publishers.py [hunk 48]
```diff
@@ -29,11 +32,17 @@ def get_course_verification_deadline(self, course):
     def serialize_seat_for_commerce_api(self, seat):
         """ Serializes a course seat product to a dict that can be further serialized to JSON. """
         stock_record = seat.stockrecords.first()
+        try:
+            enrollment_code = seat.course.enrollment_code_product
+            bulk_sku = StockRecord.objects.get(product=enrollment_code).partner_sku
```
Comment (clintonb) [inline ecommerce/courses/publishers.py]: `enrollment_code.stockrecords.first().partner_sku`

**Comment by clintonb** [comment]
An assumption is being made, but not enforced in code: A course run only has one paid seat that can be purchased in bulk. As is, this change will create an enrollment code for every single seat. This needs to be fixed, and acceptance tests need to be updated to verify.

File: ecommerce/courses/models.py (hunk 44 above)
Comment (mattdrayer) [inline ecommerce/courses/models.py:193]: Is price > 0 an acceptable filter?

Comment (clintonb) [inline ecommerce/courses/models.py:193]: Nope. The price of credit seats is greater than 0. I would create a list, `ENROLLMENT_CODE_SEAT_TYPES`, containing the seat types for which enrollment codes can be created. This should contain the verified and professional seat types.

File: ecommerce/courses/models.py (hunk 47 above)
Comment (mattdrayer) [inline ecommerce/courses/models.py:279]: I'm not sure we want to use it here, however, due to the generate_sku call below -- we would only want to generate a sku if we're in 'create' mode, but with update_or_create we would not know that ahead of time.

**Comment by clintonb** [comment]
There must be tests for the single seat assumption. The current tests do not validate real-world behavior—a course has seats of multiple types. All you need is a test that, given a course makes calls to create audit, honor, and verified seats. After these calls, there should be only one enrollment code product.

**Comment by clintonb** [comment]
👍  once test added.

**Comment by mattdrayer** [comment]
All green from Jenkins -- @vkaracic we are clear to merge this branch and can rebase 718 + 719 accordingly.

File: ecommerce/settings/base.py (hunk 39 above)
Comment (vkaracic) [inline ecommerce/settings/base.py]: There are a couple of switch names declared in the settings so I thought this is where they lived. Switches like `DISABLE_THEMING_ON_RUNTIME_SWITCH` and `PAYMENT_PROCESSOR_SWITCH_PREFIX`.

**Comment by vkaracic** [comment]
@clintonb you mentioned acceptance tests, are you saying we should write new acceptance tests for this new functionality?

**Comment by clintonb** [comment]
Yes, there should be acceptance tests.

**Comment by vkaracic** [comment]
Okay, ticket for the acceptance tests is logged: https://openedx.atlassian.net/browse/MAYN-271
